File: ipc-codegen/echo_example/rust/src/echo_client.rs
//! Echo IPC client — uses GENERATED typed client (EchoApi) over ipc-runtime.
//! Usage: echo_client --socket /tmp/echo.sock
//! Exits 0 on success, 1 on failure.

use echo_wire_compat::generated::echo_client::EchoApi;
use echo_wire_compat::generated::echo_types::{EchoInner, Fr};
use echo_wire_compat::generated::error::{IpcError, Result};
use ipc_runtime::IpcClient;

fn main() -> Result<()> {
    let args: Vec<String> = std::env::args().collect();
    let socket_path = args
        .iter()
        .position(|a| a == "--socket")
        .and_then(|i| args.get(i + 1))
        .expect("Usage: echo_client --socket <path>");

    let backend =
        IpcClient::from_path(socket_path).map_err(|e| IpcError::Backend(e.to_string()))?;
    let mut client = EchoApi::new(backend);

    // Test 1: EchoBytes
    let test_data = vec![0xDE, 0xAD, 0xBE, 0xEF, 0x42];
    let resp = client.bytes(&test_data)?;
    assert_eq!(resp.data, test_data, "EchoBytes data mismatch");
    eprintln!("echo_client(rust): EchoBytes OK");

    // Test 2: EchoFields
    let resp = client.fields(42, 999999, "hello wire compat".to_string())?;
    assert_eq!(resp.a, 42);
    assert_eq!(resp.b, 999999);
    assert_eq!(resp.name, "hello wire compat");
    eprintln!("echo_client(rust): EchoFields OK");

    // Test 3: EchoNested
    let inner = EchoInner {
        values: vec![vec![1, 2, 3], vec![4, 5]],
        flag: Some(true),
    };
    let resp = client.nested(inner.clone())?;
    assert_eq!(resp.inner.values, inner.values);
    assert_eq!(resp.inner.flag, inner.flag);
    eprintln!("echo_client(rust): EchoNested OK");

    // Test 4: EchoAliases
    let hash = Fr::from_bytes(std::array::from_fn(|i| 0x10 + i as u8));
    let second = Fr::from_bytes(std::array::from_fn(|i| 0x40 + i as u8));
    let resp = client.aliases(
        7,
        hash.clone(),
        Some(second.clone()),
        vec![hash.clone(), second.clone()],
    )?;
    assert_eq!(resp.tree_id, 7);
    assert_eq!(resp.hash, hash);
    assert_eq!(resp.maybe_hash, Some(second.clone()));
    assert_eq!(resp.hashes, vec![hash.clone(), second.clone()]);
    eprintln!("echo_client(rust): EchoAliases OK");

    // Test 5: EchoAliases with maybe_hash = None (optional-absent over live IPC)
    let resp = client.aliases(7, hash.clone(), None, vec![hash.clone()])?;
    assert_eq!(resp.maybe_hash, None);
    eprintln!("echo_client(rust): EchoAliases none OK");

    // Test 6: EchoFields with b > u32::MAX (uint64 wire encoding over live IPC)
    let big = (1u64 << 53) - 1;
    let resp = client.fields(42, big, "big".to_string())?;
    assert_eq!(resp.b, big);
    eprintln!("echo_client(rust): EchoFields u64 OK");

    // Test 7: EchoBlobs — Option<bytes> Some/None and [bytes; 2]
    let resp = client.blobs(Some(vec![0xAA, 0xBB]), [vec![1, 2, 3], vec![4]])?;
    assert_eq!(resp.maybe_data, Some(vec![0xAA, 0xBB]));
    assert_eq!(resp.parts, [vec![1, 2, 3], vec![4]]);
    let resp = client.blobs(None, [vec![], vec![9]])?;
    assert_eq!(resp.maybe_data, None);
    assert_eq!(resp.parts, [vec![], vec![9]]);
    eprintln!("echo_client(rust): EchoBlobs OK");

    // Test 8: EchoFail — server error surfaces with its message
    match client.fail("deliberate failure".to_string()) {
        Err(IpcError::Backend(message)) => {
            assert!(
                message.contains("deliberate failure"),
                "EchoFail message mismatch: {message}"
            );
        }
        other => panic!("EchoFail expected backend error, got {other:?}"),
    }
    eprintln!("echo_client(rust): EchoFail OK");

    eprintln!("echo_client(rust): all tests passed");
    Ok(())
}

File: ipc-codegen/echo_example/zig/src/echo_client.zig
/// Echo IPC client (Zig) — uses GENERATED typed client + the ipc-runtime
/// Zig binding for transport. No per-service UDS code in the example.
/// Usage: echo_client --socket /tmp/echo.sock
const std = @import("std");
const ipc_runtime = @import("ipc_runtime");
const echo_client = @import("generated/echo_client.zig");
const types = @import("generated/echo_types.zig");

fn testHash(base: u8) types.Fr {
    var hash: types.Fr = undefined;
    for (&hash, 0..) |*byte, i| {
        byte.* = base + @as(u8, @intCast(i));
    }
    return hash;
}

pub fn main() !void {
    var args = std.process.args();
    _ = args.next();
    var socket_path: ?[:0]const u8 = null;
    while (args.next()) |arg| {
        if (std.mem.eql(u8, arg, "--socket")) {
            socket_path = args.next();
        }
    }
    const path = socket_path orelse {
        std.debug.print("Usage: echo_client --socket <path>\n", .{});
        std.process.exit(1);
    };

    // Use page_allocator: the codegen-emitted client frees response buffers
    // with std.heap.page_allocator, so the runtime Client must allocate with
    // the same one.
    var backend = try ipc_runtime.Client.fromPath(std.heap.page_allocator, path);
    defer backend.deinit();

    const EchoClient = echo_client.Client(ipc_runtime.Client);
    var client = EchoClient.init(&backend);

    // Test 1: EchoBytes
    {
        const cmd = types.EchoBytes{ .data = &[_]u8{ 0xDE, 0xAD, 0xBE, 0xEF, 0x42 } };
        const resp = try client.bytes(cmd);
        if (!std.mem.eql(u8, resp.data, &[_]u8{ 0xDE, 0xAD, 0xBE, 0xEF, 0x42 })) {
            std.debug.print("echo_client(zig): EchoBytes FAIL\n", .{});
            std.process.exit(1);
        }
        std.debug.print("echo_client(zig): EchoBytes OK\n", .{});
    }

    // Test 2: EchoFields
    {
        const cmd = types.EchoFields{ .a = 42, .b = 999999, .name = "hello wire compat" };
        const resp = try client.fields(cmd);
        if (resp.a != 42 or resp.b != 999999 or !std.mem.eql(u8, resp.name, "hello wire compat")) {
            std.debug.print("echo_client(zig): EchoFields FAIL\n", .{});
            std.process.exit(1);
        }
        std.debug.print("echo_client(zig): EchoFields OK\n", .{});
    }

    // Test 3: EchoNested
    {
        const values = &[_][]const u8{ &[_]u8{ 1, 2, 3 }, &[_]u8{ 4, 5 } };
        const cmd = types.EchoNested{
            .inner = types.EchoInner{ .values = values, .flag = true },
        };
        const resp = try client.nested(cmd);
        if (resp.inner.values.len != 2) {
            std.debug.print("echo_client(zig): EchoNested FAIL\n", .{});
            std.process.exit(1);
        }
        if (resp.inner.flag != true) {
            std.debug.print("echo_client(zig): EchoNested flag FAIL\n", .{});
            std.process.exit(1);
        }
        std.debug.print("echo_client(zig): EchoNested OK\n", .{});
    }

    // Test 4: EchoAliases
    {
        const hash = testHash(0x10);
        const second = testHash(0x40);
        const hashes = &[_]types.Fr{ hash, second };
        const cmd = types.EchoAliases{
            .tree_id = 7,
            .hash = hash,
            .maybe_hash = second,
            .hashes = hashes,
        };
        const resp = try client.aliases(cmd);
        if (resp.tree_id != 7 or !std.mem.eql(u8, &resp.hash, &hash)) {
            std.debug.print("echo_client(zig): EchoAliases FAIL\n", .{});
            std.process.exit(1);
        }
        if (resp.maybe_hash == null or !std.mem.eql(u8, &resp.maybe_hash.?, &second) or resp.hashes.len != 2) {
            std.debug.print("echo_client(zig): EchoAliases optional/vector FAIL\n", .{});
            std.process.exit(1);
        }
        if (!std.mem.eql(u8, &resp.hashes[0], &hash) or !std.mem.eql(u8, &resp.hashes[1], &second)) {
            std.debug.print("echo_client(zig): EchoAliases hashes contents FAIL\n", .{});
            std.process.exit(1);
        }
        std.debug.print("echo_client(zig): EchoAliases OK\n", .{});
    }

    // Test 5: EchoAliases with maybe_hash = null (optional-absent over live IPC)
    {
        const hash = testHash(0x10);
        const cmd = types.EchoAliases{
            .tree_id = 7,
            .hash = hash,
            .maybe_hash = null,
            .hashes = &[_]types.Fr{hash},
        };
        const resp = try client.aliases(cmd);
        if (resp.maybe_hash != null) {
            std.debug.print("echo_client(zig): EchoAliases none FAIL\n", .{});
            std.process.exit(1);
        }
        std.debug.print("echo_client(zig): EchoAliases none OK\n", .{});
    }

    // Test 6: EchoFields with b > u32::MAX (uint64 wire encoding over live IPC)
    {
        const big: u64 = (1 << 53) - 1;
        const cmd = types.EchoFields{ .a = 42, .b = big, .name = "big" };
        const resp = try client.fields(cmd);
        if (resp.b != big) {
            std.debug.print("echo_client(zig): EchoFields u64 FAIL\n", .{});
            std.process.exit(1);
        }
        std.debug.print("echo_client(zig): EchoFields u64 OK\n", .{});
    }

    // Test 7: EchoBlobs — optional bytes Some/None and fixed [bytes; 2]
    {
        const cmd = types.EchoBlobs{
            .maybe_data = &[_]u8{ 0xAA, 0xBB },
            .parts = .{ &[_]u8{ 1, 2, 3 }, &[_]u8{4} },
        };
        const resp = try client.blobs(cmd);
        if (resp.maybe_data == null or !std.mem.eql(u8, resp.maybe_data.?, &[_]u8{ 0xAA, 0xBB })) {
            std.debug.print("echo_client(zig): EchoBlobs maybe_data FAIL\n", .{});
            std.process.exit(1);
        }
        if (!std.mem.eql(u8, resp.parts[0], &[_]u8{ 1, 2, 3 }) or !std.mem.eql(u8, resp.parts[1], &[_]u8{4})) {
            std.debug.print("echo_client(zig): EchoBlobs parts FAIL\n", .{});
            std.process.exit(1);
        }
        const cmd_none = types.EchoBlobs{
            .maybe_data = null,
            .parts = .{ &[_]u8{}, &[_]u8{9} },
        };
        const resp_none = try client.blobs(cmd_none);
        if (resp_none.maybe_data != null) {
            std.debug.print("echo_client(zig): EchoBlobs none FAIL\n", .{});
            std.process.exit(1);
        }
        std.debug.print("echo_client(zig): EchoBlobs OK\n", .{});
    }

    // Test 8: EchoFail — server error surfaces, message available on the client
    {
        const cmd = types.EchoFail{ .message = "deliberate failure" };
        if (client.fail(cmd)) |_| {
            std.debug.print("echo_client(zig): EchoFail FAIL (no error)\n", .{});
            std.process.exit(1);
        } else |err| {
            if (err != error.ServerError) {
                std.debug.print("echo_client(zig): EchoFail wrong error: {s}\n", .{@errorName(err)});
                std.process.exit(1);
            }
            const message = client.last_server_error orelse {
                std.debug.print("echo_client(zig): EchoFail missing message\n", .{});
                std.process.exit(1);
            };
            if (std.mem.indexOf(u8, message, "deliberate failure") == null) {
                std.debug.print("echo_client(zig): EchoFail message mismatch: {s}\n", .{message});
                std.process.exit(1);
            }
        }
        std.debug.print("echo_client(zig): EchoFail OK\n", .{});
    }

    std.debug.print("echo_client(zig): all tests passed\n", .{});
}
